sklearn/preprocessing/_encoders.py

```diff
@@ -112,7 +112,8 @@ def _transform(self, X, handle_unknown='error'):
                     # continue `The rows are marked `X_mask` and will be
                     # removed later.
                     X_mask[:, i] = valid_mask
-                    Xi = Xi.copy()
+                    # cast Xi into a np.object to handle different string types
+                    Xi = Xi.astype(np.object)
```

Comment:
Is it sufficient to cast as `self.categories_[i].dtype`?

Reply:
That can give problems in certain corner cases:

```
In [10]: categories = np.array(['A', 'B', 'C'])

In [11]: test = np.array(['A', 'Aunkonwn'])

In [12]: test.astype(categories.dtype)
Out[12]: 
array(['A', 'A'],
      dtype='<U1')
```

Basically using a numpy string dtype if you have variable length data is not a good idea anyway, and IMO you should better use object dtype to start with. But still, of course need to handle it correctly. 
We *could* also always convert any input of numpy string dtype to object before encoding / determining categories. I think the main disadvantage is that then the dtype of your `categories_` does not match the dtype of the original data.

Reply:
Ah, yes, the mask is already known, so my example was not a good one. But in general it is still a problem, as for example with

```
In [6]: categories = np.array(['A', 'B', 'CC'])

In [7]: test = np.array(['A', 'Aunkonwn'])

In [8]: test.astype(categories.dtype)
Out[8]: 
array(['A', 'Au'],
      dtype='<U2')
```

'Au' is still an unknown category